sw/device/lib/dif/dif_pinmux.c

```diff
@@ -32,9 +200,391 @@ dif_result_t dif_pinmux_output_select(const dif_pinmux_t *pinmux,
+    return kDifBadArg;
+  }
+  const dif_pinmux_lock_target_t lock_target =
+      (type == kDifPinmuxPadKindMio) ? kDifPinmuxLockTargetMioSleep
+                                     : kDifPinmuxLockTargetDioSleep;
+  bool is_locked;
+  dif_result_t result =
+      dif_pinmux_is_locked(pinmux, pad, lock_target, &is_locked);
+  if (result != kDifOk) {
+    return result;
+  }
+  if (is_locked) {
+    return kDifLocked;
+  }
+
+  // Sleep Mode Value
+  uint32_t reg_value;
+  switch (mode) {
+    case kDifPinmuxSleepModeLow:
+      reg_value = PINMUX_MIO_PAD_SLEEP_MODE_0_OUT_0_VALUE_TIE_LOW;
+      break;
+    case kDifPinmuxSleepModeHigh:
+      reg_value = PINMUX_MIO_PAD_SLEEP_MODE_0_OUT_0_VALUE_TIE_HIGH;
+      break;
+    case kDifPinmuxSleepModeHighZ:
+      reg_value = PINMUX_MIO_PAD_SLEEP_MODE_0_OUT_0_VALUE_HIGH_Z;
+      break;
+    case kDifPinmuxSleepModeKeep:
+      reg_value = PINMUX_MIO_PAD_SLEEP_MODE_0_OUT_0_VALUE_KEEP;
+      break;
+    default:
+      return kDifBadArg;
+  }
+  mmio_region_write32(pinmux->base_addr, mode_reg_offset, reg_value);
+
+  // Sleep Mode Enable
+  reg_value = bitfield_bit32_write(0, PINMUX_MIO_PAD_SLEEP_EN_0_EN_0_BIT, 1);
+  mmio_region_write32(pinmux->base_addr, en_reg_offset, reg_value);
+  return kDifOk;
+}
+
+dif_result_t dif_pinmux_pad_sleep_disable(const dif_pinmux_t *pinmux,
+                                          dif_pinmux_index_t pad,
+                                          dif_pinmux_pad_kind_t type) {
+  if (pinmux == NULL) {
+    return kDifBadArg;
+  }
+
+  ptrdiff_t en_reg_offset;
+  if (!dif_pinmux_get_reg_offset(
+          pad, type, PINMUX_MIO_PAD_SLEEP_EN_0_REG_OFFSET,
+          PINMUX_DIO_PAD_SLEEP_EN_0_REG_OFFSET, &en_reg_offset)) {
+    return kDifBadArg;
+  }
+
+  const dif_pinmux_lock_target_t lock_target =
+      (type == kDifPinmuxPadKindMio) ? kDifPinmuxLockTargetMioSleep
+                                     : kDifPinmuxLockTargetDioSleep;
+  bool is_locked;
+  dif_result_t result =
+      dif_pinmux_is_locked(pinmux, pad, lock_target, &is_locked);
+  if (result != kDifOk) {
+    return result;
+  }
+  if (is_locked) {
+    return kDifLocked;
+  }
+
+  mmio_region_write32(pinmux->base_addr, en_reg_offset, 0);
+  return kDifOk;
+}
+
+dif_result_t dif_pinmux_pad_sleep_get_state(const dif_pinmux_t *pinmux,
+                                            dif_pinmux_index_t pad,
+                                            dif_pinmux_pad_kind_t type,
+                                            bool *in_sleep_mode) {
+  if (pinmux == NULL || in_sleep_mode == NULL) {
+    return kDifBadArg;
+  }
+
+  ptrdiff_t reg_offset;
+  bitfield_bit32_index_t bit;
+  if (!dif_pinmux_get_sleep_status_bit(type, pad, &reg_offset, &bit)) {
+    return kDifBadArg;
+  }
+
+  uint32_t reg_value = mmio_region_read32(pinmux->base_addr, reg_offset);
+  *in_sleep_mode = bitfield_bit32_read(reg_value, bit);
+  return kDifOk;
+}
+
+// TODO: Should there be a faster means of clearing all bits?
```

Comment:
follow up PR?

Reply:
I simply removed the comment. This is more a question for the user, and perhaps it doesn't matter.